cmd/booster-bitswap/filters/filters.go

```diff
@@ -0,0 +1,238 @@
+			req.Header.Set("If-Modified-Since", ifModifiedSince.Format(http.TimeFormat))
+		}
+		response, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return false, nil, err
+		}
+		if response.StatusCode == http.StatusNotModified {
+			return false, nil, nil
+		}
+		if response.StatusCode < 200 && response.StatusCode > 299 {
+			bodyText, _ := io.ReadAll(response.Body)
+			return false, nil, fmt.Errorf("expected HTTP success code, got: %s, response body: %s", http.StatusText(response.StatusCode), string(bodyText))
+		}
+		return true, response.Body, nil
+	}
+}
+
+type Handler interface {
+	ParseUpdate(io.Reader) error
+	// FulfillRequest returns true if a request should be fulfilled
+	// error indicates an error in processing
+	FulfillRequest(p peer.ID, c cid.Cid) (bool, error)
+}
+
+type FilterDefinition struct {
+	Fetcher   Fetcher
+	Handler   Handler
+	CacheFile string
+}
+
+type filter struct {
+	FilterDefinition
+	lastUpdated time.Time
+}
+
+// update updates a filter from an endpoint
+func (f *filter) update() error {
+	if f.Fetcher == nil {
+		// if there's no fetcher, there's no update
+		return nil
+	}
+	fetchTime := time.Now()
+	updated, stream, err := f.Fetcher(f.lastUpdated)
+	if err != nil {
+		return fmt.Errorf("fetching endpoint: %w", err)
+
+	}
+	if !updated {
+		return nil
+	}
+	defer stream.Close()
+	// open the cache file
+	cache, err := os.OpenFile(f.CacheFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return fmt.Errorf("opening cache file: %w", err)
+	}
+	defer cache.Close()
+	forkedStream := io.TeeReader(stream, cache)
+	f.lastUpdated = fetchTime
+	err = f.Handler.ParseUpdate(forkedStream)
+	if err != nil {
+		return fmt.Errorf("parsing endpoint update: %w", err)
+	}
+	return nil
+}
+
+type MultiFilter struct {
+	cfgDir  string
+	filters []*filter
+	clock   clock.Clock
+	onTick  func()
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+func NewMultiFilterWithConfigs(cfgDir string, filterDefinitions []FilterDefinition, clock clock.Clock, onTick func()) *MultiFilter {
+	filters := make([]*filter, 0, len(filterDefinitions))
+	for _, filterDefinition := range filterDefinitions {
+		filters = append(filters, &filter{FilterDefinition: filterDefinition})
+	}
+	return &MultiFilter{
+		cfgDir:  cfgDir,
+		filters: filters,
+		clock:   clock,
+		onTick:  onTick,
+	}
+}
+
+func NewMultiFilter(
+	cfgDir string,
+	bandwidthMeasure BandwidthMeasure,
+	requestCounter RequestCounter,
+	apiFilterEndpoint string,
+	apiFilterAuth string,
+) *MultiFilter {
+	filters := []FilterDefinition{
+		{
+			CacheFile: filepath.Join(cfgDir, "denylist.json"),
+			Fetcher:   FetcherForHTTPEndpoint(BadBitsDenyList, ""),
+			Handler:   NewBlockFilter(),
+		},
+	}
+	var configFetcher Fetcher
+	if apiFilterEndpoint != "" {
+		configFetcher = FetcherForHTTPEndpoint(apiFilterEndpoint, apiFilterAuth)
+	}
+	filters = append(filters, FilterDefinition{
+		CacheFile: filepath.Join(cfgDir, "retrievalconfig.json"),
```

Comment:
do we expect cfgDir to be writable always already? is there a way for config to specify the volite directory where these caches live? (e.g. you mount in your config directory with docker or nfs, but want written files to live in /tmp)